src/osd/OSD.cc

```diff
@@ -4509,6 +4509,16 @@ bool OSD::_is_healthy()
 
   if (is_waiting_for_healthy()) {
     Mutex::Locker l(heartbeat_lock);
+    if (heartbeat_peers.empty() && osdmap->get_num_up_osds()) {
+      // We are here because there are still up peers but we failed
+      // to establish backend network connection with them.
+      // So don't mark us as healthy yet in case we'll be marked 
+      // as up again by monitor and thus may confuse all the other osds
+      // which are still maintaining a stable backend connection within
+      // each other.
+      dout(1) << "is_healthy false -- no reachable heartbeat peers" << dendl;
+      return false;
```

Comment:
@liewegas It will happen and did happen, because if this osd is on an isolated node whose backend(cluster) network has been cut off from the rest part of the cluster, it will be marked as down due to enough failure reports have been collected by monitor and will keep running in STATE_WAITING_FOR_HEALTHY state until a new osdmap is received. Later, on receiving a new CEPH_MSG_OSD_MAP message, it will try to pick up some new up osds from this new map and try to reconnect with them through the _add_heartbeat_peer() procedure, which may fail as the currently up osds are all on the other nodes and finally result its heartbeat_peer set to be empty.